ws/config.py: a replacement for the third-party modules

agalileev opened the pull request:
Hi. I made some changes for #66. `ConfigArgParse` was replaced with `argparse`, and `configfile` - with `configparser`.

After some thoughts, I decided to limit the changes and do not rework the class / function interfaces where possible. I don't know how to make tests on changes properly. Are you using unittests or something else? How do you make the "production" testing?

And yes, the syntax of the config file had to be changed to become compatible with the `configparser` module. An example can be seen in the file `examples/sample.conf`.

lahwaacz commented:
As far as I can tell, the `--site` parameter would not make sense anymore, because you can't have multiple site-specific options or even site-specific sections in the config. That's quite significant feature loss. If you can't have site-specific sections, maybe it would be better to implement at least site-specific config files.

agalileev commented:
> That's quite significant feature loss.

Is it? Eventually I have found the use case for `--config` (e.g., when you want to test the new config without replacing the one on the standard config filepath), but `--site` is looking quite useless.

1. When we deal with the config file, we work on script-specific sections. The latter inherits all options from `[DEFAULT]` section including site-specific ones. This approach is straightforward. You do not need anymore "top_level_arg" manipulations in your code which results actually to nothing (no better usability, no better config management, etc.)
2. I have thought about multiple site-specific config file management functionality, but... Do you use wiki-scripts on non-ArchWiki sites? Do you know anyone who uses it on two or more wiki projects simultaneously? I assume the answer is "no" for both questions. :-) If someone will use wiki-scripts on another site in future, he takes the config file, replaces the site-specific options, replaces script subsections and just runs the script. Simple and easy-to-use.

The old `--site` functionality cannot be restored (because `configparser` does not support such config file syntax), so I am ready to implement the functionality of multiple config files management if you are sure that this is really necessary, but it looks like an overkill to me.

lahwaacz commented:
> Do you use wiki-scripts on non-ArchWiki sites?

Yes - for example a wiki on localhost which I use for testing some stuff...

agalileev commented:
Hm, ok, I shall do it. And about testing - do you make any non-standard (I mean non-`unittest`) testing?

lahwaacz commented:
Depends on what you consider non-standard. There are "standard" unit tests for some submodules using pytest, some more involved tests which actually require internet connection to ArchWiki (they are currently disabled and probably broken) and some _very_ involved feature tests for the database synchronization part (they spawn a separate PostgreSQL process and a web server hosting a local MediaWiki instance) which never worked reliably and are also disabled. Besides that, most of the features were implemented by just running the scripts in interactive mode and observing the behaviour.

agalileev commented:
Hi. After hard painful thoughts I came up with an idea how to cross `--config`, `--site` and working with multiple config files. The two options (config and site) have been merged with the following implications:

1. There is only one hardcoded directory for configuration files now (default to `~/.config/wiki-scripts`).
2. The following naming scheme for config files are used: `$config_dir/<name>.conf`.
3. You can specify the config file you need by `--config` option: `./script.py -c <name>`.

For example, you have the following config files in your config_dir:
```
localhost.conf
very-test-config.conf
archwiki1.conf
archwiki2.conf
```
Now you can choose a config for a script just with `./script.py -c very-test-config`, `./script.py -c archwiki1`, etc.

The benefits of such approach:
1. No longer needed to specify `/long/configuration/file/path` in `--config` option. The config names are much more convenient. If you type the name which is not valid the `FileNotFoundError` will be raised (see `ws.config.argtype_configfile()` function for details).
2. Multiple configuration files can be stored in one place and simply managed.
3. Also multiple conf files for one site can be stored - just name them differently!

`ws.config` has been largely rewritten without any testing, so it potentially has the bugs. On this stage I just want your opinion about the concept.

lahwaacz commented:
Nice, I agree that merging `--config` with `--site` makes sense, but how do you determine the default `<name>`? I don't think it should be anything related to Arch wiki, as that would be inconvenient for people who want to use wiki-scripts primarily with some other wiki. I think it could be just `default`, so people would either use `default.conf` only or they could make it a symlink to their primary config file. What do you think?

I think it would also be good to interpret the value for `--config` two ways: first check if there is a `<name>.conf` in the config directory (the way you're doing now), but if it fails, check if it is a path to an existing file. Using relative paths is pretty short and convenient with tab completion. This way it would be possible to e.g. test the example config from the repo with `--config examples/sample.conf` without having to copy anything or doing something weird about the config path where `<name>` is looked up.

agalileev commented:
Ok, I have implemented two suggested changes. Unfortunately, I did not manage to stop in time and made some other changes :). I decided to split functionality of `argtype_*` between the `type=` converters and `action=` actions. It seems to me that it will be right if `type=` will do only value conversion, and `action=` will do further transformations and validity checks.

New types:
- `argtype_configfile` - converts a string to the configurations file path.
- `argtype_path` - converts a string to the absolute filesystem path.

New actions (they are registered and can be used with aliases shown below):
- `'check_path'` - check if the string is a path to a valid filesystem entry.
- `'check_dirname'` - check if the "dirname" part of a path string is an existing directory.

As for `argtype_comma_list_choices`, I brutally dealt with it. :) It looks like a nasty hack, but this task can be done much simpler. The only precondition is to use the whitespace as the option values separator:
```
# in command-line
$ ./toc.py <some options> --toc-languages en ru fr <other options>

# in configuration file
[toc]
toc-languages = en ru fr
```
See `toc.py` changes in cb26980  for more details.

It seems that I changed everything that could be changed in `ws/config.py`. :) If you have any suggestions or feature requests - I am ready to deal with them. "Please, give me that task. I hope that is the task!" (c) BlackTeaHamburger. If no then I proceed to testing, debugging and writing docs.

lahwaacz commented on ws/config.py:
```diff
@@ -11,18 +11,24 @@
 CONFIG_DIR = '~/.config/'       # Can be overriden by XDG_CONFIG_HOME
 CACHE_DIR = '~/.cache/'         # Can be overriden by XDG_CACHE_HOME
-DEFAULT_CONF = 'archwiki'       # Can be overriden with '--config' option
+DEFAULT_CONF = 'default'        # Can be overriden with '--config' option
```
Please use double quotes for all strings.

agalileev replied:
Done.

lahwaacz commented on examples/sample.conf:
```diff
@@ -1,32 +1,65 @@
+# Archwiki-related options. Replace them with the needed Mediawiki API
+# if you use wiki-scripts on anothed Wiki project.
+#
+site = ArchWiki
```
Now that the `--site` option was removed, the example should be updated.

agalileev replied:
ws/client/connection.py uses `--site` for naming cookie file:
```python
cookie_file = args.cache_dir + "/" + args.site + ".cookie"
```
I have added `add_argument("--site")` to this file directly in `set_argparser()` method.
Besides, `site` is used for configuration file interpolation:
```
cookie-file = ${cache-dir}/${site}.cookie
```
Tell me if you wish to remove this option completely.

lahwaacz commented on examples/sample.conf:
```diff
@@ -1,32 +1,65 @@
-[ArchWiki]
+#
+# Database options.
+#
```
Please use just a single-line comment.

agalileev replied:
Accepted.

lahwaacz commented on examples/sample.conf:
```diff
@@ -1,32 +1,65 @@
+[DEFAULT]
+#
+# Common options.
+#
```
This comment is pretty useless. If you want to explain what the `[DEFAULT]` section means, do it in the comment before the section.

agalileev replied:
Done.

lahwaacz commented on toc.py:
```diff
@@ -210,7 +210,7 @@ def set_argparser(argparser):
         group.add_argument("-f", "--force", action="store_true",
                 help="try to update the page even if it was last saved in the same UTC day")
-        group.add_argument("--toc-languages", default="all", type=ws.config.argtype_comma_list_choices(["all"] + lang.get_internal_tags()),
+        group.add_argument("--toc-languages", default="all", nargs="*", choices=["all"] + lang.get_internal_tags(),
```
This should be `nargs="+"`, otherwise argparse would accept even an empty list.

agalileev replied:
Yes, already fixed.

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
-        converting them back to a string representing config file contents.
+# ==============
+# Parser classes
+# ==============
```
Please remove this comment.

agalileev replied:
I adopted this style from the official `argparse` module, but ok. Done.

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
+
+        option_dict = dict(self.items(section))
+        for key, value in option_dict.items():
+            if ' ' in value:
```
This is not good - spaces do not always mean that the value should be interpreted as a list. For example, consider spaces in file paths. See https://stackoverflow.com/questions/335695/lists-in-configparser for ideas.

agalileev replied:
I figured it is a bad practice to use spaces in filenames, but ok, I'll investigate this problem.

agalileev replied:
```python
if "," in value:
    option_dict[key] = [item.strip() for item in value.split(',')]
```

Now `,` must be used as a separator in config file. In command line use `'...'` to wrap filenames with the whitespaces:
```
# in cli
./script.py --files readme.md 'hello world.c' config.py

# in config file
[script]
files = readme.md, hello world.c, config.py
```

lahwaacz replied:
Files can contain commas too. Using plain `str.split()` is not going to work, I'd consider combining it with `json` based on [this answer](https://stackoverflow.com/a/9735884).

agalileev replied:
04b94d3

Such multi-value syntax is now accepted:
```
[section]
numbers1 = [1, 2, 3, 4, 5]
numbers2 = 
    [6, 7, 8, 9, 10]
files1 = [
    "README.md",
    "hello_world.py",
    "path/to/strange/file/with spaces,and, commas"
    ]
```

Multi-line value ends with the last indented line. When `ConfigParser.fetch_section()` reads the config file, all in square brackets will be converted to a Python's list of strings.

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
+        Fetches a specific section from a config file.
+
+        :param section: section name for fetching.
+        :param to_list: defines the format of the returned value (see below).
```
Use `:param str section:` and `:param bool to_list:` since the params have an expected type.

agalileev replied:
Done.

lahwaacz replied:
I don't see the new code, you probably did not push it yet :)

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
+    def __call__(self, parser, namespace, path, option_string=None):
+        if not os.path.lexists(path)
+            raise FileNotFoundError(
+                "no such file or directory: {}".format(path)
```
I don't think this should be checked with `action="check_path"`. For example, you cannot combine it with `nargs="+"` or `action="append"` or `action="extend"`.

agalileev replied:
Actually, not every built-in action can be combined with this params (e.g., `"store_true"`). `"check_path"` and `"check_dirname"` must be used on a single argument representing a filesystem path. Do you have any other use cases in mind?

agalileev replied:
Ok, let's talk about this request. In the old-old days, there was two argtypes: `argtype_existing_dir` and `argtype_dirname_must_exist`. Both of them deal with the strings representing the filesystem path, and do the checks: the first checks if the whole path string represents the existing dir, and the second checks only the 'dirname' part of the path. I have found the opportunity to split the functionality. Now there is uniform type `argtype_path` and two actions, `'check_path'` (which now can check not only the directories, but files and links too), and `'check_dirname'`.

There is no big problem to add `nargs` support to actions code as well as some list-working functionality like `'append'` and `'extend'` do. But I need clear instructions on what kind of program behavior you expect.

And yes, I suddenly realized where I got this `os.path.abspath(os.path.expanduser(string))` construct. I took it from you. :) These two argtypes works on it. Is there any special purpose of using `abspath` or it can be wiped out?

lahwaacz replied:
So after "splitting the functionality", you have 3 objects instead of 2 to take care of. You also have limited functionality, or would have to add non-trivial amount of the same code to `check_path` and `check_dirname` to support `nargs` and `action="append"` and `action="extend"`. This is not a good direction...

Consider how you'd implement an argument taking integer values: you'd use `type=int`, because you want to convert `str` to `int` objects. And see, the `int()` "function" also does two things: it deals with the representation of integers and it checks if the argument is an integer. How would you implement an argument taking a positive integer? You would most likely create a custom function which converts the argument to `int` (and thus checks that the argument is an integer) and then checks if the value is positive. Or you could create a class such as `PositiveInt` which inherits from `int` and does the check in the `__init__` method. These two approaches are almost the same, but IMO the latter is over-complicated.

The `argtype_existing_dir` and `argtype_dirname_must_exist` functions did the same - they created some representation of a path and did some checks. The fact that the path representation uses `str` as the underlying type is irrelevant (since Python 3.4 it could use some object from [pathlib](https://docs.python.org/3/library/pathlib.html), but that is for another discussion). The command to get the path representation is so simple that there is no refactoring needed (if it was too complicated, there would be another `argtype_path` function which would just return the representation and the other argtype functions would call it before doing the checks). I really think this is a better approach than the `check_*` actions, and many other people likely think so too: the [first thing I could find on the web](https://stackoverflow.com/questions/38834378/path-to-a-directory-as-argparse-argument) uses `type` as well...

agalileev replied:
> So after "splitting the functionality", you have 3 objects instead of 2 to take care of.

You forgot about `argtype_config` which must be checked too. Action `"check_path"` can do this check, and your argtypes cannot. So it will be 3 against 3 at minimum. If one day you decide to create, for example, `argtype_file` (not for config), it also could be checked by `"check_path"`. My approach is more flexible and extensible, do you see?

I add nargs support for actions:  60acdfa 

About pathlib - I had investigated this module some days ago. I had found that the only one really usefult feature can be adopted from it - the class constructor `pathlib.PosixPath`, which can be passed to `type=` directly. All the path checking functionality must be coded by yourself anyway.

lahwaacz replied:
I did not even get to `argtype_config` yet, you can add the check just before its `return` statement to get rid of `check_path`. It's unrelated to the arguments above...

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
+        dirname = os.path.dirname(path)
+        if not os.path.isdir(dirname)
+            raise FileNotFoundError(
+                "no such directory: {}".format(dirname))
```
I don't think this should be checked with `action="check_dirname"`. For example, you cannot combine it with `nargs="+"` or `action="append"` or `action="extend"`.

lahwaacz commented on ws/db/migrations/env.py:
```diff
@@ -25,9 +25,8 @@
-    parser = ws.config.ConfigFileParser("site", "alembic")
-    conf = parser.parse(f, [])
+    parser = ws.config.ConfigParser(ws_config_path)
+    conf = parser.fetch_section('alembic', to_list=False)
```
Please always use double quotes for strings.

agalileev replied:
Ok. :)

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
     ws.logging.init(args)
-
-    # TODO: depends on ConfigArgParse, in case of argparse just log the Namespace
     logger.debug("Parsed arguments:\n" + argparser.format_values())
```
Please replace this with `logger.debug("Parsed arguments:\n{}".format(args))` (that's what the TODO note meant).

agalileev replied:
Already done (it was throwing an error).

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
+        # parameter
+        arg_defaults = {}
+
+        cache_dir = os.getenv('XDG_CACHE_HOME', os.path.expanduser(CACHE_HOME))
```
`os.path.expanduser` should be used when `CACHE_HOME` is defined.

agalileev replied:
This was a typo - `CACHE_DIR` (not `CACHE_HOME`) is defined in the start of file:

lahwaacz commented on ws/config.py:
```diff
@@ -1,180 +1,232 @@
+    # Configuration name was specified.
+    if not dirname and not ext:
+        config_dir = os.getenv('XDG_CONFIG_HOME',
+                               os.path.expanduser(CONFIG_DIR))
```
`os.path.expanduser` should be used when `CONFIG_DIR` is defined.

agalileev replied:
`CONFIG_DIR` is defined in the start of the file:
```python
CONFIG_DIR = "~/.config/"
```

lahwaacz replied:
I meant that it should be defined as `CONFIG_DIR = os.path.expanduser("~/.config/")` so that you don't have to call `expanduser` later.

agalileev replied:
I thought it is C-like mnemonic macros for the string and number literals. They [are used ](https://github.com/python/cpython/blob/3.8/Lib/argparse.py#L94)for convenience - if you want to change, for example, default config dir, you should do it only in one place (and no matter how much times it is used across the module). Function usage in global macro definition seem wrong to me.

lahwaacz replied:
They are just global variables, there is nothing resembling C macros (except maybe the uppercase style). Using `os.path.expanduser` at the definition place avoids the need to use it everywhere else, so it seems very convenient to me. I don't see any reason against doing this while you can.

agalileev replied:
I understand the difference between Python global variables and C macros, but I meant that it is used for defining the literals (and the uppercase style hints on it). We can use functions in it, but it is wrong - theoretically there is can be another usage for `"~/.config"` literal besides `os.path.expanduser` (ok, I can hardly come up with a suitable example, but _in theory_ it should work like this).

If you insist, then I will add this function, but I must warn you that this change will lead to a change in the entropy of the universe and unpredictable consequences that will be on your conscience. :-P

lahwaacz replied:
I really think that at the point where you use `CONFIG_DIR`, you shouldn't be bothered with considering whether the variable contains `~` and thus you should use `os.path.expanduser`, just because you made the assumption that global variables are string literals. If we ever changed `CONFIG_DIR` such that it does not contain `~`, there would be useless calls to `os.path.expanduser` scattered all over the place, unless somebody remembered this discussion and removed them too - but this is against the principle that changes should be done in one place only. OK, this argument is more applicable to `CACHE_DIR` rather than `CONFIG_DIR`, but still...

agalileev replied:
> unless somebody remembered this discussion and removed them too

Anyway, I do not think that to delete `os.path.expanduser()` calls will become a good idea one day. Actually, we can take path parameters in three manners:

- tilde started path
- relative path
- absolute path

Construct `os.path.abspath(os.path.expanduser(path))` is the only reliable way to convert an argument of any of the above to the valid absolute path (although in case of absolute path passed no conversion is required at all). Moreover, there is a downside: if you move `expanduser` from code to `CONFIG_DIR` assignment, the construct `os.path.abspath(CONFIG_DIR)` could lead to ambiguity. The reader will wonder every time has config path already been tilde expanded or not. "Explicit is better than implicit" (c) `import this`

In general, I slept well and I don't want to argue anymore. I'll add the call to the variable, although I will remain unconvinced. :/

agalileev replied:
Mm, in fact, the text does not contain such `os.path.abspath(os.path.expanduser())` constructions with `CONFIG_DIR`, so I will add a little explanation of what I mean:

```python
# configuration name was specified
if not dirname and not ext:
    config_dir = os.getenv("XDG_CONFIG_HOME", CONFIG_DIR)      # expanduser already moved
    path = os.path.join(config_dir,
                                  "{}/{}.conf".format(PROJECT_NAME, string))
# relative or absolute path was specified
else:
    if ext != ".conf":
        raise argparse.ArgumentTypeError(
            "config filename must end with '.conf' suffix")
        path = os.path.abspath(os.path.expanduser(string))    # expanduser can't be moved
```

On the last line we know for sure, that `string` will be tilde expanded. In the line of `config_dir = ` there is no such confidence - we must _constantly remember_ that `CONFIG_DIR` has already been tilde expanded. I cannot call it convenient. We saved a few lines of code, but have loss in clarity.

lahwaacz replied:
Converting every path to absolute is not necessary - the scripts never call `os.chdir` so relative paths will work just fine. If the user passes a relative path and you give him back an absolute path in some error message, the user will be confused.

agalileev replied:
For some reason I thought that `os.path.isdir()` function and the like takes only an absolute path as an argument. Yeah, it was a mistake.

> If the user passes a relative path and you give him back an absolute path in some error message, the user will be confused.

A reasonable remark. Now, in view of the above, it makes sense to remove `os.path.abspath()` calls where it is not strictly necessary.

agalileev commented:
Hi. I am debugging ws/config.py right now by running example scripts. And yes, I've already fixed the quotes. :) Some of your requests have already been resolved, but I will comment on every of them.

lahwaacz commented on ws/config.py:
```diff
@@ -15,33 +15,31 @@
-        super().__init__()
+    def __init__(self, configfile, **kwargs):
+        kwargs.setdefault('interpolation',
+                          configparser.ExtendedInterpolation())
```
Please don't break this on multiple lines - it looks ugly.

lahwaacz replied:
As you might have noticed, I don't really like PEP8 - I find some of its style conventions really stupid, especially the 79 chars per line limit. It's 2020! All major code editors can do soft wrapping with nice indentation. Even without visual indentation soft-wrapped code looks much better than soft-wrapping applied on hard-wrapped code.

agalileev replied:
But PEP 8... 79 characters on a line...

Upd. Oh, I wrote this comment too long. :) Ok, accepted.

lahwaacz commented on ws/config.py:
```diff
@@ -114,46 +113,46 @@ def setup(self):
+def argtype_config(string):
     """Convert `--config` argument to an absolute file path."""
     dirname = os.path.dirname(string)
     name, ext = os.path.splitext(os.path.basename(string))
```
You should not split anything. If the config name is `wiki.archlinux.org`, i.e. config path `wiki.archlinux.org.conf`, you would get `.org` in `ext`. Instead, you should just add `.conf` to the name and check if the file exists. Wait, you don't want to do "checks" in argtype functions... Well, you really should!

lahwaacz replied:
Also at this point it would be good to add some unit tests. Can you look into Pytest and try to add something?

agalileev replied:
I will write tests when all requests will be resolved and the module structure will take its final form.

agalileev replied:
ec89bae 

I left splitting but now the suffix is checked against `.conf` string explicitly. After manual testing in the interactive prompt it looks working properly.

lahwaacz commented on ws/client/connection.py:
```diff
@@ -160,6 +160,8 @@ def set_argparser(argparser):
         import ws.config
         group = argparser.add_argument_group(title="Connection parameters")
+        group.add_argument("--site", metavar="NAME",
+                help="site name used in file naming schemes")
```
This is not a connection parameter, so it should stay in `ws.config` next to the other global parameters.

agalileev commented:
I move on to writing tests. Since I haven't worked with Pytest before, this may take a while. If you have new ideas on how I can make the world a better place, just let me know. :)

agalileev commented:
Hi. I have wrote a bunch of tests for `ws.config` module. Two notes:

1. `argtype_config()` was refactored. Config filename transformation code was moved to a separate private `_get_config_filepath` function for better testing.
2. I do not know how to test `getArgParser()` and `object_from_argparser()` functions properly. They are too complex.

lahwaacz commented:
Sorry, I won't be able to review your work this week. I'll try to get to it next week, otherwise feel free to ping me if nothing happens...

agalileev commented:
No problem, I have an interesting task right now that consumes all my time. Tell me when you will ready to resume.

lahwaacz commented:
@agalileev Sorry for the delay, I'll review your work today.

agalileev commented:
Hi. I have a problem with tests after 08e6df3f. All tests for `_get_config_filepath()` are (expectedly) failed now. As I had told before, the only purpose of moving this piece of code to a separate private function was a better testing. We can test string-to-path-conversion functionality separately from file-existance-checking one. It is much more simpler than to test the relatively complex `argtype_config()` function entirely.

lahwaacz commented:
Oh, I forgot about the tests - just fixed it. I'll try to add some tests also for the `getArgParser` and `object_from_argparser` functions. I also noticed some other small problems that need more investigation, otherwise it looks pretty good.

lahwaacz commented:
Hi, I've added some tests for the `object_from_argparser` function. Unfortunately, 2 tests are failing - please see if you can fix it.

agalileev commented:
In 2e0bcbb I have fixed your changes in 08e6df3. I do now know why the absent default config must result in `args.config=None`. If the file does not exist (either default or custom) the error must be raised. Instead I add a special case `--config NONE` (NONE string is case insensitive) if you want to work without config file deliberately. Now the `argtype_config()` workflow is:

1. If `-c NONE` was specified, then `args.config=None` is set.
2. If `-c <any_string>` was specified, then corresponding file is checked for existance.
3. If no `--config` option was specified, then default config file is checked.

I mark one test as `xfail` because it is checking non-existent functionality now. A few tests for `argtype_config()` must be added, I shal do it tomorrow.

About `object_from_argparser()` tests failing. I have found, that we must parse option only ones. Otherwise anyone option (either global from `ws.config` or local from `klass.set_argparser()`) with `required=True` parameter may break the parsing. The problem is that we need to make initial parsing for `--config` option, then find a config file and make parsing again. I found a not very pretty workaround to parse `sys.argv` for `--config` manually. I am not sure that it is the best possible choice, but it looks like argparse has no suitable functionality for this case. See 8a04e49 for details.

I also add `allow_abbrev=False` to `ArgumentParser` defaults.

lahwaacz commented:
The behaviour in https://github.com/lahwaacz/wiki-scripts/commit/08e6df3f83e9a795a06689af0fdf494f5f2cba8d is intentional, please don't change it. If somebody wants to use a very simple script, e.g. `print-namespaces.py`, all that is needed is to supply a couple of options (e.g. `--api-url`) and it does not make sense to force the user to specify `--config NONE` or to create the default config file.

As for the other commit - it might be better to create a separate `ArgumentParser` instance, define just `--config` and use it for initial parsing. This should ensure that the initial parsing behaves like `argparse` - now if you pass a wrong file to `--config`, you'll get an exception traceback instead of a brief error message.

agalileev commented:
> The behaviour in 08e6df3 is intentional

This is a mistake. Imagine a situation, when the user runs a script and wants to use the default config. He does not specify `--config` option. But for some reason `default.conf` is absent or is a broken link and the user do not know about it. Normally, the error should be raised, but... nothing is happend. `args.config` just quietly set to `None` and the script proceed! Actually, it can even work without fail, produce some (possibly mistaken) output and make some (possibly mistaken) edits on Wiki. And all this time the user **has no any sign** that something goes wrong! Now imagine how much effort it will take to find the cause of the error when it will revealed?

If you want to work without config file you **must** specify this explicitly. I have seen two possible implementations for this behaviour in another applications: either `--no-config` option or a special case for standard `--config` option (like `--config NONE`).

> it might be better to create a separate ArgumentParser instance

Yes, it can be a solution. I need to make some investigation.

lahwaacz commented:
No, it is not a mistake. It should work without forcing the user to create a default config file - that just doesn't make sense. If you want, you can raise the exception when `default.conf` is a broken symlink. To explicitly skip an existing config, I'd like `--no-config` more than `--config NONE`.

agalileev commented:
I argee that creating a config as a precondition to a new user may look a bit weird, especially if he creates only one script that has no need in it. But your changes will result in undefined behavior if the `--config` option is not specified. In fact, there are two possible behaviours for every script without `--config`: to use `default.conf` or do not use any config at all. To determine which of them will take place you must check manually your `default.conf` every time you run any such script.

For example, you have three scripts to run in series:
```
./script1.py <options>        # uses default.conf
./simple-script.py <options>    # does not use config file
./script2.py <options>        # uses default.conf again
```

You need `default.conf` for `script1` and `script2`, but what will you do with it when `./simple-script` will be runned? Delete `default.conf` before and create it after? Do you really want to do such kind of work? And what if one day you delete `default.conf` for `simple-script` and forgot to restore it before `script2` is runned?

Actually, 08e6df3 kills the "default config" idea as it was planned initially. Are you afraid that user will have to type an additional `--config NONE` option? Okay, but now user must type explicit `--config archwiki` (for example) for _every_ script that intend to use config file to guarantee right behaviour.

> I'd like --no-config more than --config NONE

This can be implemented easily with `parser.add_mutually_exclusive_group()` for `--config` and `--no-config` options, but we need to work out the final concept of `--config` option first.

agalileev commented:
I think, we can use a more verbose error message as a possible solution, for example:

```
file does not exist or is a broken link: '<path_to_config>'. Create the config file or use '--no-config' option.
```

This a) explains the new user what he have to do; b) do not produce dangerous behaviour.

lahwaacz commented:
There is no undefined behaviour in the case I intend: if `--config` is not specified, the behaviour depends on whether the default config exists or not, but the behaviour is well defined in either case.

Let me rephrase again. The intention is to have a shared config for all scripts, which supports sections for per-script options. So to specify options only for specific scripts, they need to be in a specific section in the config file - there is no danger of unintentionally changing the behaviour of other scripts. If you put them in the `[DEFAULT]` section, I think the section name is good enough to indicate that such options affect all scripts. This is more or less the same behaviour as there is now, I don't want to change that.

agalileev commented:
> the behaviour depends on whether the default config exists or not

But this is exactly a problem! Look:

```
$ ./script.py --cache-dir <some_dir> --api-url <some_url>
```

A question: "Does this script use `defautl.conf`?"
An answer: "_I do not know. May be yes, and may be not._ It depends on whether the default config exists or not. May be there are another options in `klass.set_argparser()` that must be read from default config, but if there is no `default.conf` on filesystem they wil be set to default values, which are possibly mistaken for this run. We have no guarrantee that the script will behave as we intend because we do not now is there `default.conf` on filesystem or not".

Another example:

```
$ ./script1.py --cache-dir <...> --api-url <...>
$ ./script2.py --cache-dir <...> --api-url <...>
```

A question: "Can you determine with this two strings (without checking the filesystem for a file), what script is working with `default.conf`, and what is working without any config?"
An answer: "No, I cannot"

Do you see? This is what I mean "undefined behaviour". If we have two separate logics on "`--config` is not specified" case then one command can do two different things depending on `default.conf` presence. This is the source of big problems in the future. You will have to constantly check for the file if you have two types of scripts and want to guarantee no mistakes.

And now look for an alternative:

```
$ ./script1.py --no-config --cache-dir <...> --api-url <...>
$ ./script2.py --cache-dir <...> --api-url <...>
```

Here we can say with confidence:
- script1.py is working without the config file; `--cache-dir` and `--api-url` are the only two options in use.
- script2.py is working with the `default.conf` file; `--cache-dir` and `--api-url` is taken from the command line, and if there are any options set in `klass.set_argparser()`, they will be taken from the config file. (And if the `default.conf` does not exist the error will be raised to warn the user).

This is what I can call "well-defined behaviour" :)

lahwaacz commented:
If the user can't remember if their config file exists or not, they can't expect anything from the configuration system. If you want a nicer way to check which values apply for a given script invokation, you can try to reimplement [ConfigArgParse's print_values function](https://github.com/lahwaacz/ConfigArgParse/blob/0bdc6544f780d2dc5fa48b0aeedfa68edb8b4650/configargparse.py#L570-L599) and bind it to an option such as `--print-values`. But I don't find this feature necessary.

As for "constantly checking for the file": of course the user will never do this, they will use the defaults first (which will either succeed or fail), then they might pass a cli option or create a config file for the script to _change the default_. But this is _not_ a problem, because __if you configure the script in the proper config file section, it will not affect other scripts in any way__. Again, it is the user's responsibility to write the config in such a way that they don't get surprised in the future.

Please remember that this PR is about replacing third-party modules, not about major reimplementation of the configuration system.

agalileev commented:
Okaaay, I roll changes back to 08e6df3. Just for the record — this is a mistake :(

I need more time to fix "args.config parsing", but for now all tests are passing.

In our discussion it [sounded indistinct](https://github.com/lahwaacz/wiki-scripts/pull/69#issuecomment-691910756), so I will clarify — do we need in `--no-config` option implemented or this will be a _major reimplementation of the configuration system_ too?

lahwaacz commented:
> In our discussion it [sounded indistinct](https://github.com/lahwaacz/wiki-scripts/pull/69#issuecomment-691910756), so I will clarify — do we need in `--no-config` option implemented or this will be a _major reimplementation of the configuration system_ too?

I don't consider this to be _major_, so I'll leave it up to you.

agalileev commented:
Hi. I have tried some ideas how to handle args.config parsing without an error. The core problem was that if we use two fully separate parsers than the options only from the first of them gets in `--help` output — but if we use the single parser, any option with `required=True` can break the parsing.

I have found that the problem can be solved if to inject config parser into the main parser with the `parents=` parameter. I have no idea how (and why) this works, but it works... The tradeoff is that `getArgParser()` has the interface changed. It returns a tuple of two now. The first element is the main parser, and the second is a new "parent" parser with the only `--config`/`--no-config` options group.

If the script uses only `object_from_argparser()` it will work without changes. If the script uses `getArgParser()` it must be changed. The string

```python
argparser = ws.config.getArgParser()
```

must be replaced with

```python
argparser, _ = ws.config.getArgParser()
```

I have changed `examples/parse-config.py` script for example.

May be I have something missed, but all tests are passed.

lahwaacz commented:
Hi, good job. Actually we don't need to use `parents=`, only `add_help=False` for the auxiliary parser with the config arguments.

I also moved some things around and split the config file parsing from `object_from_argparser` into a separate function `parse_args`, because I realized that there are scripts which use plain `getArgParser` without `object_from_argparser` and the config file would not be parsed by those scripts.

There is still (at least) one behavioral difference, though: config files should be allowed to contain unrecognized arguments/variables, but unrecognized arguments specified on the command line should be reported as errors. I've added two test cases for this, one of them fails.

agalileev commented:
I have added the check for unrecognized options in cli_args. But there is one case which is not covered by this logic: when the multiple short options are joined together with a single `-` prefix. `argparse` manages such cases well, but checking these options manually will result in a big piece of code.

Moreover, such options can produse a separate error message. Consider parser with only defined `-a` option which is not taken explicit value from the command line (e.g., `action="store_true"`). Parsing `args, remainder = ap.parse_known_args(["-abc"])` produces an error:

```
error: argument -a: ignored explicit argument 'bc'
```

It does not perceive `b` and `c` as options at all and tries to treat `bc` as an explicit value for `-a` option.

agalileev commented:
Theoretically, the problem can be solved with two-step parsing. We can parse `config_args` with the `parse_known_args()` which do not check for unrecognized options, and the parse `cli_args` with `parse_args()` (which do). It will be "a right way". But here we will face with the old problem: any `required=True` option with high probability breaks parsing and the whole script.

I am not be able to investigate this problem until next Tuesday because I have some jobs I must to do (including monthly syncing of AW articles). After that I will be ready to fight this PR until death, its or mine :).

lahwaacz commented:
> But there is one case which is not covered by this logic: when the multiple short options are joined together with a single - prefix. argparse manages such cases well, but checking these options manually will result in a big piece of code.

I think it can be handled quite easily, since short options are not allowed in config files, so any short option remaining after the parsing means an error. It is just a matter of figuring out the right error message (and checking that short options in the config file are actually not possible).

> I am not be able to investigate this problem until next Tuesday

Take your time, there is no rush - I'm quite busy these days too...

agalileev commented:
Hi. It looks like I understand how to work with multioptions. Consider such case:

```python
_, remainder = ap.parse_known_args(["-abc"], namespace=args)
```

1. When `argparse` sees multioption block, he tries to recognize the separate options it consists. If the first of them  (`-a`) was defined earlier `argparse` makes parsing as usually, with assignments or raising errors. We cannot do something here because this is `argparse`'s internal logic.
2. If `-a` was not defined `argparse` places the whole block to `remainder` as unrecognized. Here we can catch it with `if item.startswith("-") and item in cli_args` statement and make some error handling.

I have also added the check for short options in a config file and some tests.

agalileev commented:
Hi, Jakub. Do you have any update for this pull request? I will have a lot of free time this month, so if you have any  ideas or suggestions I can work on them.

lahwaacz commented:
Hi, sorry I did not find time for this in the last month. I'll try to look at it at the weekend.

agalileev commented:
Ok, look. The main purpose that I intended with this PR was to get practice in Python programming on real project, and more or less, the goal has been achieved. If you are not very enthusiastic at the thought of merging this PR (for whatever reason), then there is no need to continue. We can simply close it to save our time for other jobs. :)

lahwaacz commented:
No, just closing the PR would be a waste of our already spent effort. It's not about enthusiasm, it is a good idea and I actually like your changes and the general direction this PR took, but I simply do not have as much free time these days as I had in the summer.

I think that the code is mostly finished, but before merging we should write some brief documentation and/or a migration guide for other users (they are not many but we should still be nice to them). Or at least make a "beta" release or ask them for testing it or something like that. What I mean to say here is that the last step is more than just clicking the merge button.

agalileev commented:
Ok, then I shall wait. Tell me if I can do something to help.

lahwaacz commented:
This took much longer than I'd like to admit, but it is finally merged! I apologize for letting you wait so long...

I've released [version 1.3](https://github.com/lahwaacz/wiki-scripts/releases/tag/1.3) just before the merge, so we can polish the details and documentation for the next version (which will be probably 2.0).

agalileev commented:
Oh my!.. This day has finally come. It looks like I shall drink vodka for a month or two...
